### PR TITLE
CMP: support securely getting root CA cert update

### PR DESCRIPTION
**Pull request by DDvO**
This pertains to the CMP API, app, doc, and tests.

The CMP client may send a `genm` request message with infoType `id-it-rootCaCert` (CLI option `-infotype rootCaCert`)
for retrieving any available root CA cert update. The server responds with a `genp` message with infoType `id-it-rootCaKeyUpdate` possibly containing the new certificate and transition certs`newWithOld` and `oldWithNew`.

If provided , the client verifies it using the transition certificate `newWithOld` etc. as defined in RFC 4210 section 4.4. The usual certificate verify function does not directly support this, but is tweaked using a custom chain building callback function. 

The clients saves on success the new cert in a file named by the `-newwithnew` option, as well as
the transition certificates (as far as provided by the server and file names given to the client).

This PR also includes related small improvements as separate commits:
* CMP app and API doc: add note on critical server auth on receiving trust anchor certs
* CMP doc: various text improvements, mostly on PBM vs. MAC-based protection
* rename the new function `OSSL_CMP_get{,1}_caCerts` (still possible for version 3.2) and improve `OSSL_CMP_exec_certreq.pod`
* APPS `load_key_certs_crls()`: improve error string macro `FAIL_NAME`

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
Would some of the documentation updates be applicable to stable branches?

**Comment by DDvO** [comment]
Yes. I will provide respektive backports.

**Comment by DDvO** [comment]
> Yes. I will provide respective backports.

Done in #21138

**Comment by DDvO** [comment]
Thank you @t8m for your yesterday's quick review of #21116.
Can you please also resume reviewing this PR soon?
This feature is high on our (Siemens) CMP priority list.

**Comment by DDvO** [comment]
Rebased to fix a simple merge conflict in the docs.

Can we please get this PR forward soon, such that it will make it into 3.2?

**Review by t8m — CHANGES_REQUESTED** [review]
One minor nit

File: test/recipes/80-test_cmp_http.t [hunk 1]
```diff
@@ -28,6 +28,8 @@ plan skip_all => "These tests are not supported in a no-cmp build"
     if disabled("cmp");
 plan skip_all => "These tests are not supported in a no-ec build"
     if disabled("ec");
+plan skip_all => "These tests are not supported in a no-ecx build"
+    if disabled("ecx"); # EDDSA test key material in Mock/newWithNew.pem
```
[the remark this answers is not in the document]
Comment (DDvO) [reply test/recipes/80-test_cmp_http.t:32]: > disabled("ec") implies disabled("ecx")

Yes, but according to my experience not the other way round.
Yesterday, after rebasing, I got a CI failure for `no-ecx` on a test case involving the mentioned EDDSA cert:
https://github.com/openssl/openssl/actions/runs/5516732133/jobs/10058437759?pr=21129
So I added `if disabled("ecx")` and the failure disappeared.
Thus I believe both exclusions are needed, at least for the current state of the exclusion mechanism.

[the remark this answers is not in the document]
Comment (t8m) [reply test/recipes/80-test_cmp_http.t:32]: No, they aren't. You just leave `disabled("ecx")` and remove the `disabled("ec")`.

[the remark this answers is not in the document]
Comment (DDvO) [reply test/recipes/80-test_cmp_http.t]: Indeed, due to the implication already present.
Done.

**Review by t8m — APPROVED** [review]

File: doc/man3/OSSL_CMP_exec_certreq.pod [hunk 2]
```diff
@@ -114,27 +119,59 @@ typically return them as an indication that the certificate was already revoked.
...
+given root CA certificate I<oldWithOld> and verifies it as far as possible.
+See RFC 4210 section 4.4 for details.
+On success it assigns to I<*newWithNew> the root certificate received.
+As far as the I<newWithOld> and I<oldWithNew> output parameters are not NULL,
+it assigns to them the corresponding transition certificates.
+NULL means that the respective certificate was not provided by the server.
+All certificates obtained this way must be freed by the caller.
+
+B<WARNING:>
+The I<newWithNew> certificate is meant to be certificate that will be trusted.
```
Comment (paulidale) [inline doc/man3/OSSL_CMP_exec_certreq.pod:151]: meant to be **a** certificate

File: doc/man1/openssl-cmp.pod.in [hunk 3]
```diff
@@ -669,6 +676,43 @@ The file where to save the list of CA certificates contained in the caPubs field
...
+If on success no such cert was received, this file (if present) is deleted
+to indicate that the requested root CA certificate update is not available.
+
+Any received newWithNew certificate is verified
+using any received newWithOld certificate as untrusted intermediate certificate
+and the certificate provided with B<-oldwithold> as the (only) trust anchor,
+or if not provided, using the certificates given with the B<-trusted> option.
+
+B<WARNING:>
+The newWithNew certificate is meant to be certificate that will be trusted.
```
[the remark this answers is not in the document]
Comment (DDvO) [reply doc/man1/openssl-cmp.pod.in]: Oops, fixed, also in `OSSL_CMP_exec_certreq.pod`.

File: doc/man1/openssl-cmp.pod.in [hunk 4]
```diff
@@ -1125,6 +1185,17 @@ only affect the certificate verification enabled via the B<-out_trusted> option.
 
 =head1 NOTES
 
+When a client obtains from a CMP server CA certificates that it is going to
```
[the remark this answers is not in the document]
Comment (DDvO) [reply doc/man1/openssl-cmp.pod.in]: done

File: doc/man3/OSSL_CMP_exec_certreq.pod [hunk 5]
```diff
@@ -114,27 +119,59 @@ typically return them as an indication that the certificate was already revoked.
...
-NULL means that no CA certificate is available at the server.
+NULL output means that no CA certificates were provided by the server.
+
+OSSL_CMP_get1_rootCaKeyUpdate() uses a genm request message
+with infoType rootCaCert to obtain from the CMP server referenced by I<ctx>
+in a genp response message with infoType rootCaKeyUpdate any update of the
+given root CA certificate I<oldWithOld> and verifies it as far as possible.
+See RFC 4210 section 4.4 for details.
+On success it assigns to I<*newWithNew> the root certificate received.
+As far as the I<newWithOld> and I<oldWithNew> output parameters are not NULL,
```
[the remark this answers is not in the document]
Comment (DDvO) [reply doc/man3/OSSL_CMP_exec_certreq.pod]: done

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
Merged.

**Pull request closed without merging**